### PR TITLE
Display comment time in the session's time zone #4488

### LINKED ISSUE #4488: Display comment time in the session's time zone
Currently, comment timestamp is in the format: Mon, 13 Jul 2015, 08:47:18 UTC
We need to do this using the session time zone.

### PR DESCRIPTION
**Pull request by Mynk96**
Fixes #4488

### CODE REVIEW

**Comment by Mynk96** [comment]
Ready For Review!

File: src/main/java/teammates/storage/search/FeedbackResponseCommentSearchDocument.java [hunk 1]
```diff
@@ -300,6 +300,7 @@ public static FeedbackResponseCommentSearchResultBundle fromResults(
                 frcDb.deleteDocument(comment);
                 continue;
             }
+            bundle.sessionTimeZone = session.getTimeZone();
```
Comment (leeyimin) [inline src/main/java/teammates/storage/search/FeedbackResponseCommentSearchDocument.java]: The time zone of the bundle would be the time zone of the session of the last document in the filtered results. If there are multiple sessions in the filtered result, the time zone of certain comment rows may be wrong.

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 2]
```diff
@@ -73,6 +73,50 @@
...
+        mapDoubleToId("8.75", "Australia/Eucla");
+        mapDoubleToId("9.0", "Asia/Tokyo");
+        mapDoubleToId("9.5", "Australia/Adelaide");
+        mapDoubleToId("10.0", "Australia/Canberra");
+        mapDoubleToId("10.5", "Australia/Lord_Howe");
+        mapDoubleToId("11.0", "Pacific/Noumea");
+        mapDoubleToId("12.0", "Pacific/Auckland");
+        mapDoubleToId("12.75", "Pacific/Chatham");
+        mapDoubleToId("13.0", "Pacific/Tongatapu");
+        mapDoubleToId("14.0", "Pacific/Kiritimati");
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: This may not be the best way to translate the double to a TimeZone object. I tested with some of the time zones which corresponded to the US cities in the list and DST was taken in account (eg. UTC -0800 shows up as -0700), which may be unexpected for users. 

https://docs.oracle.com/javase/7/docs/api/java/util/TimeZone.html 
Using the custom time zone ID may be more appropriate here, and it can be generated by code as well.

**Comment by Mynk96** [comment]
@leeyimin Can you please have a look at build error!

File: src/test/resources/pages/instructorFeedbackResultsAddComment.html [hunk 3]
```diff
@@ -358,7 +358,7 @@
                               <li class="list-group-item list-group-item-warning" id="responseCommentRow-0-1-0-1-1">
                                 <div id="commentBar-0-1-0-1-1">
                                   <span class="text-muted">
-                                    From: Teammates Test [${datetime.now}]
+                                    From: Teammates Test [Fri, 30 Jun 2017, 05:24 AM +0800]
```
Comment (leeyimin) [inline src/test/resources/pages/instructorFeedbackResultsAddComment.html]: You need to update the `replaceUnpredictableValuesWithPlaceholders` method in HtmlHelper. As the time zone of the comment time has changed, the HtmlHelper is unable to replace occurrences of current time correctly. One solution I can think of now is to loop through all possible time zones and replace all possible current time representations with the placeholder. Would be good if you can find a better solution :joy:

Comment (whipermr5) [inline src/test/resources/pages/instructorFeedbackResultsAddComment.html]: Ideal case: detect the time zone and replace with `${datetime.now.utc+8}`

**Comment by Mynk96** [comment]
Ready for Review!

**Comment by Mynk96** [comment]
Ready for Review!

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 4]
```diff
@@ -458,4 +459,15 @@ public static String convertToRequiredFormat(String date, String hour, String mi
 
     }
 
+    public static String getCustomIdOfTimeZone(double sessionTimeZone) {
+        String sign = sessionTimeZone > 0 ? "+" : "";
+        int hours = (int) sessionTimeZone;
+        int minutes = (int) ((sessionTimeZone - Math.floor(sessionTimeZone)) * 60);
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: The formula is not correct for negative time zones which are not `-*` or `-*.5`. 
For example, `-4.25` should return `"GMT-04:15"` but `"GMT-04:45"` instead. 
While there are no such time zones in the list of time zones supported, it is preferred if the logic of this method works for all time zones, not just the ones supported.

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 5]
```diff
@@ -215,20 +215,21 @@ public static String formatTime12H(Date date) {
...
-    public static String formatDateTimeForComments(Date date) {
+    public static String formatDateTimeForComments(Date date, double sessionTimeZone) {
         if (date == null) {
             return "";
         }
         SimpleDateFormat sdf = null;
         Calendar c = Calendar.getInstance(SystemParams.TIME_ZONE);
         c.setTime(date);
+        String timeZone = getCustomIdOfTimeZone(sessionTimeZone);
         if (c.get(Calendar.HOUR_OF_DAY) == 12 && c.get(Calendar.MINUTE) == 0) {
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: You may need to update this line as well. While the calendar hour may be 12, the formatted hour may not be 12, which makes the added `" NOON"` string unsuitable. On this note, you should add more tests to TimeHelperTest to ensure your method is working correctly.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 6]
```diff
@@ -492,7 +492,7 @@ public static String injectContextDependentValuesForTest(String content) {
                                StringHelper.truncateLongId(TestProperties.TEST_ADMIN_ACCOUNT))
                       .replace("<!-- nexthour.date -->", TimeHelper.formatDate(TimeHelper.getNextHour()))
                       .replace("<!-- now.datetime -->", TimeHelper.formatTime12H(now))
-                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now));
+                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now, 8));
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: This change does not seem necessary for this PR. I would like to know why this line was changed.

File: src/main/java/teammates/common/util/TimeHelper.java (hunk 5 above)
Comment (Mynk96) [inline src/main/java/teammates/common/util/TimeHelper.java]: I am not getting why we are showing `12 NOON`. We can just show `12 am` or `12 pm`.

Comment (whipermr5) [inline src/main/java/teammates/common/util/TimeHelper.java]: From a quick [search](https://github.com/TEAMMATES/teammates/issues?utf8=✓&q=is%3Aissue%20noon%20) in the issue tracker: #4667

Comment (Mynk96) [inline src/main/java/teammates/common/util/TimeHelper.java]: `Calendar.HOUR_OF_DAY` stores time in 24-hour format. If it is 12 means it is `12 NOON` only. 
Apologies, If I misunderstood anything. Can you please elaborate a bit more.

Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: Suppose the original time in UTC was `1st Jan 2016 1200 UTC NOON` , and the desired time zone is +1 GMT, your method would output something like `1st Jan 1300 +0100 NOON`, which would not be what is expected.

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 6 above)
Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: This line is added to change `[${datetime.now}]` to `[${datetime.now} +0800]` in test outputs.

**Comment by Mynk96** [comment]
@leeyimin Please Review!

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 6 above)
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Is there any reason +8 GMT was used specifically?

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 7]
```diff
@@ -215,20 +215,22 @@ public static String formatTime12H(Date date) {
...
         c.setTime(date);
+        sdf = new SimpleDateFormat("EEE, dd MMM yyyy, hh:mm a Z");
+        sdf.setTimeZone(timeZone);
         if (c.get(Calendar.HOUR_OF_DAY) == 12 && c.get(Calendar.MINUTE) == 0) {
-            sdf = new SimpleDateFormat("EEE, dd MMM yyyy, hh:mm");
-            sdf.setTimeZone(SystemParams.TIME_ZONE);
-            return sdf.format(date) + " NOON UTC";
+            sdf = new SimpleDateFormat("EEE, dd MMM yyyy, hh:mm Z");
+            sdf.setTimeZone(timeZone);
+            return sdf.format(date) + " NOON";
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: I believe "NOON" should come before the timezone, you can look up https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html for how you can add text in the pattern.

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 8]
```diff
@@ -215,20 +215,22 @@ public static String formatTime12H(Date date) {
...
         }
-        sdf = new SimpleDateFormat("EEE, dd MMM yyyy, hh:mm a zzz");
-        sdf.setTimeZone(SystemParams.TIME_ZONE);
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: Sorry, I did not notice that the two patterns were different. You may want to revert the change to move the related lines up.

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 9]
```diff
@@ -458,4 +460,16 @@ public static String convertToRequiredFormat(String date, String hour, String mi
 
     }
 
+    public static TimeZone getTimeZoneFromDoubleOffset(double sessionTimeZone) {
+        String sign = sessionTimeZone > 0 ? "+" : "";
+        int hours = (int) sessionTimeZone;
+        int minutes = (int) ((Math.abs(sessionTimeZone) - Math.floor(Math.abs(sessionTimeZone))) * 60);
+        String offset = Integer.toString(hours);
+        if (minutes != 0) {
+            offset = String.format("%s:%s", Integer.toString(hours), Integer.toString(minutes));
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: There are better ways to format this string, see https://docs.oracle.com/javase/7/docs/api/java/util/Formatter.html#syntax.
One way to simplify the code is to use `String.format("GMT%+03d:%02d", hours, minutes)`

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 10]
```diff
@@ -36,7 +36,7 @@
     private static final String REGEX_BLOB_KEY = "(encoded_gs_key:)?[a-zA-Z0-9-_]{10,}";
     private static final String REGEX_QUESTION_ID = "[a-zA-Z0-9-_]{40,}";
     private static final String REGEX_COMMENT_ID = "[0-9]{16}";
-    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M( UTC)?";
+    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M( -+[0-9])?";
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: The issue is that if the UTC date differs from the test case time zone date, an assertion error will occur. It also does not properly ascertain that the correct time zone is used. Refer to previous comment from @whipermr5 on how it can be improved?

File: src/test/java/teammates/test/cases/util/TimeHelperTest.java [hunk 11]
```diff
@@ -203,8 +203,27 @@ public void testEndOfYearDates() {
...
+
+        cal.clear();
+        cal.set(2015, 10, 30, 4, 0, 0);
+        date = cal.getTime();
+        assertEquals("Mon, 30 Nov 2015, 12:00 +0800 NOON", TimeHelper.formatDateTimeForComments(date, 8));
+
+        cal.clear();
+        cal.set(2015, 10, 30, 4, 0, 0);
+        date = cal.getTime();
+        assertEquals("Mon, 30 Nov 2015, 04:00 PM +1200", TimeHelper.formatDateTimeForComments(date, 12));
```
Comment (leeyimin) [inline src/test/java/teammates/test/cases/util/TimeHelperTest.java]: Fixed.

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 10 above)
Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: > Ideal case: detect the time zone and replace with ${datetime.now.utc+8}
As mentioned, I think currently we are doing the same only.With `new Date()` we are getting the date in system's time zone and replacing it with `${datetime.now}+0800`.

Comment (whipermr5) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: The code will need to do more than just a regex replace. It'll have to parse the timestamp, convert it to UTC, and check whether it's the same as the current time. If it is, replace with `{datetime.now+X}` where X is the correct time zone.

Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Suppose the UTC date is 11/07/2017 while the GMT+8 date is 12/07/2017. This will cause a html assertion error although the GMT +8 date was correct. We want to avoid these false negative results. It would be good if the regex used is specific to the timezone so that we will be able to assert that the eventually replaced result is displayed in the correct time zone.

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 12]
```diff
@@ -458,4 +460,11 @@ public static String convertToRequiredFormat(String date, String hour, String mi
 
     }
 
+    public static TimeZone getTimeZoneFromDoubleOffset(double sessionTimeZone) {
+        int hours = (int) sessionTimeZone;
+        int minutes = (int) ((Math.abs(sessionTimeZone) - Math.floor(Math.abs(sessionTimeZone))) * 60);
+        String customId = TimeZone.getTimeZone(String.format("GMT%+03d:%02d", hours, minutes)).getID();
+        return TimeZone.getTimeZone(customId);
```
Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: Return the timezone in line 466?

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 13]
```diff
@@ -36,9 +39,10 @@
     private static final String REGEX_BLOB_KEY = "(encoded_gs_key:)?[a-zA-Z0-9-_]{10,}";
     private static final String REGEX_QUESTION_ID = "[a-zA-Z0-9-_]{40,}";
     private static final String REGEX_COMMENT_ID = "[0-9]{16}";
-    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M( UTC)?";
+    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M?";
     private static final String REGEX_ADMIN_INSTITUTE_FOOTER = ".*?";
     private static final String REGEX_SESSION_TOKEN = REGEX_UPPERCASE_HEXADECIMAL_CHAR_32;
+    private static final String REGEX_TIMEZONE_OFFSET = "(([\\+][0-9][0-9][0-9][0-9])|([\\-][0-9][0-9][0-9][0-9]))";
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: The regex is long and quite hard to read. `"UTC([+-]\\d{4})"` could be one better way to write this.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 14]
```diff
@@ -36,9 +39,10 @@
     private static final String REGEX_BLOB_KEY = "(encoded_gs_key:)?[a-zA-Z0-9-_]{10,}";
     private static final String REGEX_QUESTION_ID = "[a-zA-Z0-9-_]{40,}";
     private static final String REGEX_COMMENT_ID = "[0-9]{16}";
-    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M( UTC)?";
+    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M?";
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: `([AP]M)?` or `([AP]M|NOON)` ?

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 15]
```diff
@@ -385,6 +389,7 @@ private static String suppressVariationsInInjectedValues(String content) {
     private static String replaceUnpredictableValuesWithPlaceholders(String content) {
         Date now = new Date();
         SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy, ");
+        sdf.setTimeZone(getTimeZone(content));
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: May want to add comment about how this assumes that only one timezone would be used on the page.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 16]
```diff
@@ -436,7 +441,7 @@ private static String replaceUnpredictableValuesWithPlaceholders(String content)
                       .replaceAll("plainCommentText-" + REGEX_COMMENT_ID, "plainCommentText-\\${comment\\.id}")
                       // date of next hour (datepicker's date is generated based on next hour's date)
                       .replace(dateOfNextHour, "${date.nexthour}")
-                      // date/time now e.g [Thu, 07 May 2015, 07:52 PM] or [Thu, 07 May 2015, 07:52 PM UTC]
+                      // date/time now e.g [Thu, 07 May 2015, 07:52 PM] or [Thu, 07 May 2015, 07:52 PM +0000]
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Remove the second example since `+0000` is not replaced.

**Comment by Mynk96** [comment]
@leeyimin On a different note, currently comment time format is `Mon, 30 Nov 2015, 04:00 PM UTC+1200` (till the last commit) instead of using `UTC+0800` can we show offset in `General Time Zone` i.e. `GMT + 08:00`? I think that will make it look more coherent and clear.

**Comment by leeyimin** [comment]
@Mynk96 I would think it's better to use UTC in the display since UTC was used when asking users to select the timezone, unless you want to change all usage of UTC to GMT. Hmm, why do you feel use of GMT is better than UTC as well?

**Comment by whipermr5** [comment]
> General Time Zone i.e. GMT + 08:00

@Mynk96 Btw, GMT stands for Greenwich Mean Time

**Comment by Mynk96** [comment]
@whipermr5 I was just giving a sample format how we can show comment time.

**Comment by Mynk96** [comment]
@leeyimin Taking back my suggestion. `UTC` would be better as we have users using `teammates` around the globe.

**Comment by whipermr5** [comment]
@Mynk96 I see; apologies 😅

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 17]
```diff
@@ -215,20 +215,22 @@ public static String formatTime12H(Date date) {
...
-            return sdf.format(date) + " NOON UTC";
+            sdf = new SimpleDateFormat("EEE, dd MMM yyyy, hh:mm 'NOON' 'UTC'Z");
+            sdf.setTimeZone(timeZone);
+            return sdf.format(date);
         }
-        sdf = new SimpleDateFormat("EEE, dd MMM yyyy, hh:mm a zzz");
-        sdf.setTimeZone(SystemParams.TIME_ZONE);
+        sdf = new SimpleDateFormat("EEE, dd MMM yyyy, hh:mm a 'UTC'Z");
+        sdf.setTimeZone(timeZone);
         return sdf.format(date);
```
Comment (Mynk96) [inline src/main/java/teammates/common/util/TimeHelper.java]: We don't need to change this as `Calendar.HOUR_OF_DAY` stores time in `24-hour-clock`, so 12 here means 12 NOON only.

Comment (leeyimin) [inline src/main/java/teammates/common/util/TimeHelper.java]: The code quality here can be improved since the only difference is "'NOON'" or "a".

Comment (Mynk96) [inline src/main/java/teammates/common/util/TimeHelper.java]: Yes, Got it!

**Comment by Mynk96** [comment]
@leeyimin Please Review!

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 18]
```diff
@@ -385,6 +389,8 @@ private static String suppressVariationsInInjectedValues(String content) {
     private static String replaceUnpredictableValuesWithPlaceholders(String content) {
         Date now = new Date();
         SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy, ");
+        //A page can have only one time zone as we cannot have session with two different time zones.
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: A page can have two time zones since instructor comments can show comments from different courses and sessions. You may want to correct your comment.

Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: You should only mention that the time zone of expected `dateTimeNow` is taken from the content.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 19]
```diff
@@ -492,7 +498,18 @@ public static String injectContextDependentValuesForTest(String content) {
...
-                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now));
+                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now, 0));
+    }
+
+    private static TimeZone getTimeZone(String content) {
+        Pattern pattern = Pattern.compile(REGEX_TIMEZONE_OFFSET);
+        Matcher matcher = pattern.matcher(content);
+        //set default timezone offset.
+        String timeZoneOffset = "+0000";
+        while (matcher.find()) {
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Why is `while` used here?

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 20]
```diff
@@ -36,9 +39,10 @@
     private static final String REGEX_BLOB_KEY = "(encoded_gs_key:)?[a-zA-Z0-9-_]{10,}";
     private static final String REGEX_QUESTION_ID = "[a-zA-Z0-9-_]{40,}";
     private static final String REGEX_COMMENT_ID = "[0-9]{16}";
-    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M( UTC)?";
+    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M?|NOON";
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: The regex should be improved, it matches "12:00 A", "NOON" but not "12:00 NOON".

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 21]
```diff
@@ -502,14 +502,19 @@ public static String injectContextDependentValuesForTest(String content) {
     }
 
     private static TimeZone getTimeZone(String content) {
+        //Searches for last String of pattern "UTC+0800" in the content.
         Pattern pattern = Pattern.compile(REGEX_TIMEZONE_OFFSET);
         Matcher matcher = pattern.matcher(content);
-        //set default timezone offset.
+        //set default time zone offset.
         String timeZoneOffset = "+0000";
         while (matcher.find()) {
```
Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: As the content will have comments from single time zone, I am finding the last word with Pattern 'UTC+XXXX` and extracting time zone offset from it.

Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Why is the last word with `UTC+XXXX` used, not the first one?

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 22]
```diff
@@ -385,6 +389,8 @@ private static String suppressVariationsInInjectedValues(String content) {
     private static String replaceUnpredictableValuesWithPlaceholders(String content) {
         Date now = new Date();
         SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy, ");
+        //Get session's time zone from content.
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: You should mention in the comment that this assumes that the page only shows one timezone as it is possible for a page to show multiple time zones (eg. InstructorSearchPage) and this method would not work for such a page.

Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Add a space after `//` and use lower case for `get`.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 23]
```diff
@@ -492,7 +498,23 @@ public static String injectContextDependentValuesForTest(String content) {
                                StringHelper.truncateLongId(TestProperties.TEST_ADMIN_ACCOUNT))
                       .replace("<!-- nexthour.date -->", TimeHelper.formatDate(TimeHelper.getNextHour()))
                       .replace("<!-- now.datetime -->", TimeHelper.formatTime12H(now))
-                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now));
+                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now, 0));
+    }
+
+    private static TimeZone getTimeZone(String content) {
+        //Searches for last String of pattern "UTC+0800" in the content.
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: This comment does not seem relevant. You should not be searching for "UTC+8000" in particular.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 24]
```diff
@@ -36,9 +39,10 @@
     private static final String REGEX_BLOB_KEY = "(encoded_gs_key:)?[a-zA-Z0-9-_]{10,}";
     private static final String REGEX_QUESTION_ID = "[a-zA-Z0-9-_]{40,}";
     private static final String REGEX_COMMENT_ID = "[0-9]{16}";
-    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] [AP]M( UTC)?";
+    private static final String REGEX_DISPLAY_TIME = "(0[0-9]|1[0-2]):[0-5][0-9] ([AP]M?|NOON)";
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: I believe the `?` after `M` is unnecessary?

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 25]
```diff
@@ -492,7 +498,23 @@ public static String injectContextDependentValuesForTest(String content) {
...
+        //set default time zone offset.
+        String timeZoneOffset = "+0000";
+        while (matcher.find()) {
+            timeZoneOffset = matcher.group(1);
+        }
+        return TimeZone.getTimeZone("GMT" + removeUtc(timeZoneOffset));
+    }
+
+    private static String removeUtc(String timeZoneOffset) {
+        return timeZoneOffset.substring(0, 2);
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: The method can be improved, it should still work for time zones which are not integer offsets (eg. +0330).

Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: This method is just converting string from `UTC+xxxx` to `xxxx`.

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 21 above)
Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: because using pattern `+xxxx` it also matches string such as responseCommentRow-`1234`567812345678 in the content file.

Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: I believe you can simply use the first word of `UTC+XXXX`. Is there any reason you must use the last one?

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 25 above)
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: I believe you are taking the first 2 characters of `timeZoneOffset`.

Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: No, I am searching for the last string with a pattern as `UTC+xxxx` and then extracting `+xxxx` from it.For example, `timeZoneOffset` is `UTC-0330` then this will return `-0330`.

Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: By matching group 1, and not group 0, you are already extracting `+xxxx`. Please read up on methods you use as well, `String.substring(int beginIndex, int endIndex)` returns the first 2 characters. It does not remove the first three characters.

Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Apologies. I have improved the method.

File: src/test/java/teammates/test/driver/HtmlHelper.java (hunk 21 above)
Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Yes, we can also use the first one.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 26]
```diff
@@ -492,7 +498,19 @@ public static String injectContextDependentValuesForTest(String content) {
                                StringHelper.truncateLongId(TestProperties.TEST_ADMIN_ACCOUNT))
                       .replace("<!-- nexthour.date -->", TimeHelper.formatDate(TimeHelper.getNextHour()))
                       .replace("<!-- now.datetime -->", TimeHelper.formatTime12H(now))
-                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now));
+                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now, 0));
+    }
+
+    private static TimeZone getTimeZone(String content) {
+        // searches for last String of pattern "UTC+xxxx" in the content.
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Update your comment.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 27]
```diff
@@ -385,6 +389,8 @@ private static String suppressVariationsInInjectedValues(String content) {
     private static String replaceUnpredictableValuesWithPlaceholders(String content) {
         Date now = new Date();
         SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy, ");
+        // get session's time zone from content.
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Mention that this assumes only the page only shows time in one time zones, there are pages which can show multiple time zones (eg. InstructorSearchPage) and this method may not work on such pages.

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 28]
```diff
@@ -492,7 +498,19 @@ public static String injectContextDependentValuesForTest(String content) {
...
                       .replace("<!-- now.datetime -->", TimeHelper.formatTime12H(now))
-                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now));
+                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now, 0));
+    }
+
+    private static TimeZone getTimeZone(String content) {
+        // searches for last String of pattern "UTC+xxxx" in the content.
+        Pattern pattern = Pattern.compile(REGEX_TIMEZONE_OFFSET);
+        Matcher matcher = pattern.matcher(content);
+        //set default time zone offset.
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: space after `//`.

**Comment by Mynk96** [comment]
@leeyimin Please Review!

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 29]
```diff
@@ -385,6 +389,9 @@ private static String suppressVariationsInInjectedValues(String content) {
     private static String replaceUnpredictableValuesWithPlaceholders(String content) {
         Date now = new Date();
         SimpleDateFormat sdf = new SimpleDateFormat("EEE, dd MMM yyyy, ");
+        // get session's time zone from content (content here includes pages only with one time zone).
```
Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: `content here includes pages only with one time zone` is not accurate. Remove this part?

Comment (Mynk96) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Should I replace it with something or `// get session's time zone from content` is fine?

Comment (leeyimin) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: Yes, that would be fine.

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 30]
```diff
@@ -458,4 +457,10 @@ public static String convertToRequiredFormat(String date, String hour, String mi
 
     }
 
+    public static TimeZone getTimeZoneFromDoubleOffset(double sessionTimeZone) {
+        int hours = (int) sessionTimeZone;
+        int minutes = (int) ((Math.abs(sessionTimeZone) - Math.floor(Math.abs(sessionTimeZone))) * 60);
+        return TimeZone.getTimeZone(String.format("GMT%+03d:%02d", hours, minutes));
```
Comment (HirdayGupta) [inline src/main/java/teammates/common/util/TimeHelper.java]: Can change from GMT to UTC?

Comment (whipermr5) [inline src/main/java/teammates/common/util/TimeHelper.java]: The Java `TimeZone` object only accepts `GMT`.

Comment (HirdayGupta) [inline src/main/java/teammates/common/util/TimeHelper.java]: Ohh I see. Its okay then I guess. :)

**Comment by Timezone1** [comment]
I think this is not good way to translate the double object of time zone.You can check [time zone](https://stopwatch360.com/timezone/) how to work for more you can see the source code.
